crates/circuit/src/circuit_data.rs

```diff
@@ -995,6 +996,24 @@ impl CircuitData {
         self.param_table.clear();
     }
 
+    /// Counts the number of times each operation is used in the circuit.
+    ///
+    /// # Parameters
+    /// - `self` - A mutable reference to the CircuitData struct.
+    ///
+    /// # Returns
+    /// An IndexMap containing the operation names as keys and their respective counts as values.
+    pub fn count_ops(&mut self) -> IndexMap<String, usize, ::ahash::RandomState> {
+        let mut ops_count: IndexMap<String, usize, ::ahash::RandomState> = IndexMap::default();
+        for instruction in &self.data {
+            *ops_count
+                .entry(instruction.op.view().name().to_string())
+                .or_insert(0) += 1;
+        }
+        ops_count.par_sort_by(|_k1, v1, _k2, v2| v2.cmp(v1));
+        ops_count
+    }
```

Comment:
I took a quick look locally and if you make these changes you can use an `&str` here instead of `String`. This should actually have a noticeable impact on performance because a `String` is an owned object and with `to_string` you'll be copying it (`String` is essentially a `Vec<u8>`). While if you can use `&str` that will avoid any copies and use references for all the strings; at least until we return to Python.

```suggestion
    pub fn count_ops(&self) -> IndexMap<&str, usize, ::ahash::RandomState> {
        let mut ops_count: IndexMap<&str, usize, ::ahash::RandomState> = IndexMap::default();
        for instruction in &self.data {
            *ops_count.entry(instruction.op.name()).or_insert(0) += 1;
        }
        ops_count.par_sort_by(|_k1, v1, _k2, v2| v2.cmp(v1));
        ops_count
    }
```